soroban-env-host/src/host.rs

```diff
@@ -1204,6 +1273,49 @@ impl VmCallerCheckedEnv for Host {
     type VmUserState = Host;
     type Error = HostError;
 
+    // This function is somewhat subtle.
+    //
+    // It exists to allow the client of the (VmCaller)CheckedEnv interface(s) to
+    // essentially _reject_ an error returned by one of the Result-returning
+    // methods on the trait, choosing to panic instead. But doing so in some way
+    // that the trait defines, rather than calling panic in the client.
+    //
+    // The only "client" we expect to _do_ this is the `impl Env for CheckedEnv`
+    // definition, in checked_env.rs, which itself is only used for testing
+    // native contracts, adapting `CheckedEnv` to the non-error-returning `Env`
+    // interface that contracts expect.
+    //
+    // When such a "rejected error" occurs, we do panic, but only after checking
+    // to see if we're in a `TestContract` invocation, and if so storing the
+    // error's Status value in that frame, such that `Host::call_n` above can
+    // recover the Status when it _catches_ the panic and converts it back to an
+    // error.
+    //
+    // It might seem like we ought to `std::panic::panic_any(e)` here, making
+    // the panic carry a `HostError` or `Status` and catching it by dynamic type
+    // inspection in the `call_n` catch logic. The reason we don't do so is that
+    // `panic_any` will not provide a nice printable value to the `PanicInfo`,
+    // it constructs, so when/if the panic makes it to a top-level printout it
+    // will display a relatively ugly message like "thread panicked at Box<dyn
+    // Any>" to stderr, when it is much more useful to the user if we have it
+    // print the result of HostError::Debug, with its glorious status code,
+    // site-of-origin backtrace and debug log.
```

Comment:
💯 Agreed, this seems like the right trade-off for us to be making to not use that mechanism.